woocommerce/woocommerce
# Add migration to move orders from wp_posts to custom tables

### All Submissions:






### Changes proposed in this Pull Request:



Closes #31766.

To add migrations, we add a general migration class that targets any combination of an entity table along with it's metadata table, described in the `MetaToCustomTableMigrator` class. We then use objects of this class in `WPPostToCOTMigrator` class providing the necessary configuration for each custom table.

### How to test the changes in this Pull Request:

~This PR should be tested along with #32052, #32053 and #32054. For code review, I have split it into four PRs, but for testing, it should be fine to only test the `cot/migration-operational-data` as it has all the commits.~

We have now merged the separate PRs into this one, so all code is here and reviewing and testing just this one should be good enough.

1. Checkout the `cot/migration-operational-data` branch (see PR #32054).
2. Make sure that you have `{prefix}_wc_orders`, `{prefix}_wc_order_addresses` and `{prefix}_wc_order_operational_data` created by following steps from #31811.
3. Run this snippet:
```php

namespace Automattic\WooCommerce\DataBase\Migrations\CustomOrderTable;

$done = false;
$migrator = new WPPostToCOTMigrator();
delete_option( 'wc_cot_migration' );
$done = $migrator->process_next_migration_batch( 500 );
```

Check contents of `{prefix}_wc_orders`, `{prefix}_wc_order_addresses` and `{prefix}_wc_order_operational_data`  table and verify that first 500 order records are migrated.

5. After the above commands you can also optionally run this to migrate all the records:
```php

while( ! $done ) {
	$done = $migrator->process_next_migration_batch( 500 );
}
```

Note the PR has many `TODO` comments, which will be handled by different PRs as the resolution for them is not clear at this time.

### Other information:




### Changelog entry

> Enhancement - Add migration for moving data from wp_posts to custom order table.

### FOR PR REVIEWER ONLY:
